### PR TITLE
JAVA FLE2 work: AutoEncryptOptions and ClientEncryptOptions work

### PR DESCRIPTION
**Pull request by rozza**

### CODE REVIEW

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java [hunk 1]
```diff
@@ -333,6 +316,18 @@ public void close() {
         }
     }
 
+    private void configure(final Supplier<Boolean> successfulConfiguration) {
```
Comment (jyemin) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:319]: It's not necessarily successful. Rename to just `configuration`?

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java [hunk 2]
```diff
@@ -333,6 +316,18 @@ public void close() {
...
+        if (!successfulConfiguration.get()) {
+            throwExceptionFromStatus();
+        }
+    }
+
+    private void configure(final Supplier<Boolean> successfulConfiguration, final mongocrypt_ctx_t context) {
```
Comment (jyemin) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:325]: It's not necessarily successful. Rename to just `configuration`?

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java [hunk 3]
```diff
@@ -277,34 +265,32 @@ public MongoCryptContext createExplicitEncryptionContext(final BsonDocument docu
...
-                    MongoCryptContextImpl.throwExceptionFromStatus(context);
-                }
+                configure(() -> mongocrypt_ctx_setopt_key_alt_name(context, keyAltNameBinaryHolder.getBinary()), context);
             }
         }
 
-        success = mongocrypt_ctx_setopt_algorithm(context, new cstring(options.getAlgorithm()), -1);
-        if (!success) {
-            MongoCryptContextImpl.throwExceptionFromStatus(context);
+        if (INDEXED.equalsIgnoreCase(options.getAlgorithm())) {
```
Comment (jyemin) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:278]: I'm not sure I understand what's going on in this if/else block.  Why are we sometimes calling  mongocrypt_ctx_setopt_index_type and someting calling  mongocrypt_ctx_setopt_algorithm for the same option?

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java [hunk 4]
```diff
@@ -19,13 +19,44 @@
...
+import java.util.Objects;
+
 /**
  * Options for explicit encryption.
  */
 public class MongoExplicitEncryptOptions {
     private final BsonBinary keyId;
     private final String keyAltName;
     private final String algorithm;
+    private final Long contentFactor;
```
Comment (jyemin) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java:31]: contentFactor -> contentionFactor

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java [hunk 5]
```diff
@@ -114,17 +178,46 @@ public String getAlgorithm() {
...
 
     @Override
     public String toString() {
         return "MongoExplicitEncryptOptions{" +
                 "keyId=" + keyId +
-                ", keyAltName=" + keyAltName +
-                ", algorithm='" + algorithm + "'}";
+                ", keyAltName='" + keyAltName + '\'' +
+                ", algorithm='" + algorithm + '\'' +
+                ", contentFactor=" + contentFactor +
```
Comment (jyemin) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java:220]: Add queryType?

**Review by jyemin — CHANGES_REQUESTED** [review]

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java (hunk 4 above)
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java]: Done

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java (hunk 2 above)
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:325]: Renamed to `successSupplier`

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java (hunk 1 above)
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:319]: Renamed to successSupplier

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java (hunk 3 above)
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:278]: Followed the logic from: https://github.com/kevinAlbs/mongo-go-driver/pull/3/files#diff-e2829099000d774cf77f3bed96c37b4afa44ce09fedddc4b58ebeb91019442beR183-R197

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java (hunk 5 above)
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoExplicitEncryptOptions.java]: Done

**Review by jyemin — APPROVED** [review]
LGTM

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java (hunk 3 above)
Comment (kevinAlbs) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:278]: If no algorithm is set, does `getAlgorithm()` return null? Is that OK in this comparison?

**Review by kevinAlbs — APPROVED** [review]
LGTM with one question about null handling of algorithm.

**Review by markbenvenuto — APPROVED** [review]

**Comment by rozza** [comment]
Manually rebased and merged.

**Pull request closed without merging**